Sources/RealModule/ApproximateEquality.swift

```diff
@@ -0,0 +1,154 @@
+      to: other,
+      absoluteTolerance: relativeTolerance * Magnitude.leastNormalMagnitude,
+      relativeTolerance: relativeTolerance
+    )
+  }
+  
+  /// Compares `self` and `other` for approximate equality with a specified tolerances.
+  ///
+  /// `true` if `self` and `other` are equal, or if they are finite and either
+  /// ```
+  /// (self - other).magnitude ≤ absoluteTolerance
+  /// ```
+  /// or
+  /// ```
+  /// (self - other).magnitude ≤ relativeTolerance * scale
+  /// ```
+  /// where `scale` is `max(self.magnitude, other.magnitude)`.
+  ///
+  /// Mathematical Properties:
+  /// -
+  /// - `isApproximatelyEqual(to:absoluteTolerance:relativeTolerance:)`
+  ///   is _reflexive_ for non-exceptional values (such as NaN).
+  ///
+  /// - `isApproximatelyEqual(to:absoluteTolerance:relativeTolerance:)`
+  ///   is _symmetric_.
+  ///
+  /// - `isApproximatelyEqual(to:absoluteTolerance:relativeTolerance:)`
+  ///   is __not__ _transitive_. Because of this, approximately equality is
+  ///   __not an equivalence relation__, even when restricted to non-exceptional values.
+  ///
+  /// - For any point `a`, the set of values that compare approximately equal to `a` is _convex_.
+  ///   (Under the assumption that `norm` implements a valid norm, which cannot be checked
+  ///   by this function.)
+  ///
+  /// - Parameters:
+  ///
+  ///   - other: The value to which `self` is compared.
+  ///
+  ///   - absoluteTolerance: The absolute tolerance to use in the comparison.
+  ///
+  ///     This value should be non-negative and finite.
+  ///     This constraint on is only checked in debug builds, because a mathematically
+  ///     well-defined result exists for any tolerance, even one out of range.
+  ///
+  ///   - relativeTolerance: The relative tolerance to use in the comparison.
+  ///     Defaults to zero.
+  ///
+  ///     This value should be non-negative and less than or equal to 1.
+  ///     This constraint on is only checked in debug builds, because a mathematically
+  ///     well-defined result exists for any tolerance, even one out of range.
+  ///
+  ///   - norm: The norm to use for the comparison. Defaults to the `.magnitude`
+  ///     property.
+  ///
+  ///     When porting code or working with specialized algorithms, it may
+  ///     be desirable to override the norm. For example, if we wanted to test if
+  ///     a complex value was inside a circle of radius 0.001 centered at (1 + 0i), we
+  ///     could use:
+  ///     ```
+  ///     z.isApproximatelyEqual(
+  ///       to: 1,
+  ///       absoluteTolerance: 0.001,
+  ///       norm: \.length
+  ///     )
+  ///     ```
+  ///     (if we used the default, we would be testing if `z` were inside a square region
+  ///     instead.)
+  @inlinable @inline(__always)
+  public func isApproximatelyEqual(
+    to other: Self,
+    absoluteTolerance: Magnitude,
+    relativeTolerance: Magnitude = 0,
+    norm: (Self) -> Magnitude = { $0.magnitude }
+  ) -> Bool {
```

Comment:
I wonder if taken together `isApproximateEqual(to:relativeTolerance:)` and `isApproximateEqual(to:absoluteTolerance:relativeTolerance:norm:)` are a minor violation of the API guideline:

> Prefer to locate parameters with defaults toward the end of the parameter list. Parameters without defaults are usually more essential to the semantics of a method, and provide a stable initial pattern of use where methods are invoked.

because it's not providing "a stable initial pattern"? 🤷‍♂️

Reply:
I pondered that, but I think it's OK because you can _also_ use `isApproximatelyEqual(to:absoluteTolerance:)`. I don't think there's clear guidance in the API guideline for when a:, b:, and a:b: are all allowed, nor can there be.

Reply:
Is there any good reason Julia `isapprox` lists `rtol` before `atol`?

Reply:
I think it's just that they want that to be the privileged option, but they don't have an explicit `isapprox(a,b;rtol)` overload like we do.

Reply:
That makes sense. I don't have strong opinion here, but IIUC we could achieve a similar effect by adding another overload (here ignoring `norm`):

```swift
extension Real {
    func isApproximatelyEqual(to other: Self, relativeTolerance: Magnitude = Magnitude.ulpOfOne.squareRoot())
    func isApproximatelyEqual(to other: Self, absoluteTolerance: Magnitude)
    func isApproximatelyEqual(to other: Self, relativeTolerance: Magnitude, absoluteTolerance: Magnitude)
}
```

Is `norm` only useful if you have an `absoluteTolerance`, or should we do something like:

```swift
extension Real {
    func isApproximatelyEqual(to other: Self, relativeTolerance: Magnitude = Magnitude.ulpOfOne.squareRoot(), norm: (Self) -> Magnitude = { $0.magnitude })
    func isApproximatelyEqual(to other: Self, absoluteTolerance: Magnitude, norm: (Self) -> Magnitude = { $0.magnitude })
    func isApproximatelyEqual(to other: Self, relativeTolerance: Magnitude, absoluteTolerance: Magnitude, norm: (Self) -> Magnitude = { $0.magnitude })
}
```